### PR TITLE
disallow MD5 in CertificateBuilder and CertificateSigningRequestBuilder

### LINKED ISSUE #3737: x509_builder.sign(ec_key, hashes.MD5()) produces an assertion error from OpenSSL
`'error:100C508A:elliptic curve routines:pkey_ec_ctrl:invalid digest type'`

It should error with a clean python exception, not an `InternalError`.

### PR DESCRIPTION
**Pull request by reaperhulk**
fixes #3737

### CODE REVIEW

**Comment by alex** [comment]
Sorry, I thought we were going to handle the error on ecdsa, as is this
breaks rsa.

**Comment by reaperhulk** [comment]
Do we need to preserve compatibility there? Is it useful outside of generating test cases? And can those test cases not be generated out of band via the openssl CLI?

**Comment by alex** [comment]
idk, I thought that was the point of deprecating all of MD5, to avoid these questions.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -7,6 +7,10 @@ Changelog
 .. note:: This version is not yet released and is under active development.
 
 * **BACKWARDS INCOMPATIBLE:** Support for Python 3.3 has been dropped.
+* **BACKWARDS INCOMPATIBLE:** The ``MD5`` hash algorithm can't be used
+  for signing with
+  :class:`~cryptography.x509.CertificateBuilder` and
+  :class:`~cryptography.x509.CertificateSigningRequestBuilder`.
```
Comment (alex) [inline CHANGELOG.rst]: This should be updated, maybe even removed, MD5 + EC _never_ worked!

File: tests/test_x509.py [hunk 2]
```diff
@@ -2623,6 +2687,41 @@ def test_sign_invalid_hash_algorithm(self, backend):
...
+
+        builder = x509.CertificateSigningRequestBuilder().subject_name(
+            x509.Name([
+                x509.NameAttribute(NameOID.ORGANIZATION_NAME, u'PyCA'),
+            ])
+        )
+        request = builder.sign(private_key, hashes.MD5(), backend)
+        assert isinstance(request.signature_hash_algorithm, hashes.MD5)
+
+    @pytest.mark.requires_backend_interface(interface=RSABackend)
```
Comment (alex) [inline tests/test_x509.py]: Wrong interface

File: tests/test_x509.py [hunk 3]
```diff
@@ -2623,6 +2687,41 @@ def test_sign_invalid_hash_algorithm(self, backend):
...
+                x509.NameAttribute(NameOID.ORGANIZATION_NAME, u'PyCA'),
+            ])
+        )
+        with pytest.raises(ValueError):
+            builder.sign(private_key, hashes.MD5(), backend)
+
+    @pytest.mark.requires_backend_interface(interface=EllipticCurveBackend)
+    def test_sign_ec_with_md5(self, backend):
+        _skip_curve_unsupported(backend, ec.SECP256R1())
+        private_key = ec.generate_private_key(ec.SECP256R1(), backend)
```
Comment (alex) [inline tests/test_x509.py]: Use a fixture for performance.

File: tests/test_x509.py [hunk 4]
```diff
@@ -1933,6 +1933,70 @@ def test_sign_with_unsupported_hash(self, backend):
...
+            datetime.datetime(2032, 1, 1, 12, 1)
+        )
+        with pytest.raises(ValueError):
+            builder.sign(private_key, hashes.MD5(), backend)
+
+    @pytest.mark.requires_backend_interface(interface=EllipticCurveBackend)
+    @pytest.mark.requires_backend_interface(interface=X509Backend)
+    def test_sign_ec_with_md5(self, backend):
+        _skip_curve_unsupported(backend, ec.SECP256R1())
+        private_key = ec.generate_private_key(ec.SECP256R1(), backend)
```
Comment (alex) [inline tests/test_x509.py]: fixture

**Comment by alex** [comment]
Does this impact CRLs as well?

File: CHANGELOG.rst (hunk 1 above)
Comment (reaperhulk) [inline CHANGELOG.rst]: Yeah we don't need this with the updated approach.

**Pull request merged**